src/components/Excerpt.tsx

```diff
@@ -0,0 +1,224 @@
+  linkStyle: Record<string, string>;
+};
+
+/**
+ * An optional toggle link at the bottom of an excerpt which controls whether
+ * it is expanded or collapsed.
+ */
+function InlineControls({
+  isCollapsed,
+  setCollapsed,
+  linkStyle,
+}: InlineControlsProps) {
+  return (
+    <div
+      className={classnames(
+        // Position these controls at the bottom right of the excerpt
+        'absolute block right-0 bottom-0',
+        // Give extra width for larger tap target and gradient fade
+        // Fade transparent-to-white left-to-right to make the toggle
+        // control text (More/Less) more readable above other text.
+        // This gradient is implemented to-left to take advantage of Tailwind's
+        // automatic to-transparent calculation: this avoids Safari's problem
+        // with transparents in gradients:
+        // https://bugs.webkit.org/show_bug.cgi?id=150940
+        // https://tailwindcss.com/docs/gradient-color-stops#fading-to-transparent
+        'w-20 bg-gradient-to-l from-white',
+      )}
+    >
+      <div className="flex justify-end">
+        <LinkButton
+          variant="text"
+          onClick={() => setCollapsed(!isCollapsed)}
+          expanded={!isCollapsed}
+          title="Toggle visibility of full excerpt text"
+          style={linkStyle}
+          underline="always"
+          inline
+        >
+          {isCollapsed ? 'More' : 'Less'}
+        </LinkButton>
+      </div>
+    </div>
+  );
+}
+
+const noop = () => {};
+
+export type ExcerptProps = {
+  children?: ComponentChildren;
+
+  /**
+   * If `true`, the excerpt provides internal controls to expand and collapse
+   * the content. If `false`, the caller sets the collapsed state via the
+   * `collapse` prop.  When using inline controls, the excerpt is initially
+   * collapsed.
+   */
+  inlineControls?: boolean;
+
+  /**
+   * If the content should be truncated if its height exceeds
+   * `collapsedHeight + overflowThreshold`. This prop is only used if
+   * `inlineControls` is false.
+   */
+  collapse?: boolean;
+
+  /**
+   * Maximum height of the container, in pixels, when it is collapsed.
+   */
+  collapsedHeight: number;
+
+  /**
+   * An additional margin of pixels by which the content height can exceed
+   * `collapsedHeight` before it becomes collapsible.
+   */
+  overflowThreshold?: number;
+
+  /**
+   * Called when the content height exceeds or falls below
+   * `collapsedHeight + overflowThreshold`.
+   */
+  onCollapsibleChanged?: (isCollapsible: boolean) => void;
+
+  /**
+   * When `inlineControls` is `false`, this function is called when the user
+   * requests to expand the content by clicking a zone at the bottom of the
+   * container.
+   */
+  onToggleCollapsed?: (collapsed: boolean) => void;
+
+  /**
+   * Additional styles to pass to the inline controls element.
+   * Ignored if inlineControls is false
+   */
+  inlineControlsLinkStyle?: Record<string, string>;
```

Comment:
This could use the `JSX.CSSProperties` type to make the supported values more obvious.